File: common/src/main/java/com/hartwig/actin/TreatmentDatabase.java
package com.hartwig.actin;

import java.util.Map;

import com.hartwig.actin.clinical.datamodel.treatment.Drug;
import com.hartwig.actin.clinical.datamodel.treatment.Treatment;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class TreatmentDatabase {

    @NotNull
    private final Map<String, Drug> drugsByName;
    @NotNull
    private final Map<String, Treatment> treatmentsByName;

    public TreatmentDatabase(@NotNull Map<String, Drug> drugsByName, @NotNull Map<String, Treatment> treatmentsByName) {
        this.drugsByName = drugsByName;
        this.treatmentsByName = treatmentsByName;
    }

    @Nullable
    public Treatment findTreatmentByName(@NotNull String treatmentName) {
        return treatmentsByName.get(treatmentName.toLowerCase());
    }

    @Nullable
    public Drug findDrugByName(@NotNull String drugName) {
        return drugsByName.get(drugName.toLowerCase());
    }
}

File: common/src/main/kotlin/com/hartwig/actin/TreatmentDatabase.kt
package com.hartwig.actin

import com.hartwig.actin.datamodel.clinical.treatment.Drug
import com.hartwig.actin.datamodel.clinical.treatment.DrugTreatment
import com.hartwig.actin.datamodel.clinical.treatment.Treatment

class TreatmentDatabase(val drugsByName: Map<String, Drug>, val treatmentsByName: Map<String, Treatment>) {

    fun drugs() = drugsByName.values.toSet()

    fun treatments() = treatmentsByName.values.toSet()

    fun findTreatmentByName(treatmentName: String): Treatment? {
        return treatmentsByName[treatmentName.replace(" ", "_").lowercase()]
    }

    fun findDrugByName(drugName: String): Drug? {
        return drugsByName[drugName.replace(" ", "_").lowercase()]
    }

    fun findDrugByAtcName(atcName: String): Drug? {
        return findDrugByName(atcName.lowercase().replace(" and ", "_").split(", ").first())
    }

    fun findDrugTreatmentByDrugs(drugs: Set<Drug>): Treatment? {
        return treatments().filterIsInstance<DrugTreatment>().firstOrNull { it.drugs == drugs }
    }
}
